### ISSUE TITLE
Domain model package structure needs improvement

### ISSUE THREAD

**Issue by mike1813** [Spyderisk/system-modeller#103]
To deploy a new domain model or update an existing one, we create a zipfile containing the necessary artefacts and upload it to system-modeller via the 'KnowledgeBase Manager' page.

The zipfile is created in two ways:

- by the CI pipeline in the Spyderisk project for the domain model
- manually by domain model developers wishing to test their changes

If we want more people to be involved in domain model development, we need the process of creating the zipped bundle to be as simple as possible.

At present, the structure of the zipfile is fixed:

- a copy of the domain model NQ file which must be called 'domain.nq'
- a second copy of the same NQ file which must be called '\<domaingraph\>.nq'
- a folder containing icon image files (currently we use SVG format, though I think some other formats are supported)
- a file which must be called `icon-mapping.json` that provides a mapping from assertible asset classes to icon filenames

The last file starts with two elements:

- "ontology": "\<domaingraph\>",
- "graph": "`http://it-innovation.soton.ac.uk/ontologies/trustworthiness/<domaingraph>`",

The second of these specifies the graph URI where the domain model will be stored in the system-modeller's triple store.

When a domain modeller creates this bundle manually, the NQ and JSON files are generated using [csv2nq](https://github.com/Spyderisk/domain-csv2nq), using command line options to specify their filenames. During development, one normally needs to distinguish the new versions from the previous one, so these filenames cannot be fixed.

This leaves two problems for the domain modeller when inserting these files into a zipfile:

- it is necessary to copy the csv2nq outputs to create copies with the required filenames
- the NQ file has to be copied twice

The second of these is not just irritating, it is embarrassing. It exposes the fact that the domain model updating procedure contains some kludges that really should be resolvable. I believe that should be fixed as a priority, so that the zipfile only needs to contain one NQ file.

The first problem is irritating but not embarrassing. The idea that system-modeller expects a fixed filename isn't unreasonable, so if we can get down to one NQ file and one JSON file, it would be OK to used fixed filenames like 'domain.nq' and 'icon-mapping.json'. The filenames should not depend on \<domaingraph\>, as to make it depend on something else is a complication a domain modeller could do without. The best solution would be to allow any filenames but require the NQ and JSON filenames to end with '.nq' and '.json', respectively.

So - I propose we implement one of three possible solutions:

- Ideal: the zip bundle must contain an icons folder plus one NQ and one JSON file, with arbitrary filenames ending '.nq' and '.json', respectively.
- Next best: the zip bundle must contain an icons folder plus one NQ and one JSON file, called 'domain.nq' and 'icon-mapping.json', respectively.
- Worst but still an improvement on what we have: the zip bundle must contain an icons folder plus one NQ and one JSON file, whose names are constrained in a way that depends on the \<domaingraph\>.

@danshearer, @scp93ch : we should not seek to involve more people in domain model development until one of these has been implemented.

**Comment by kenmeacham** [comment]
I've been investigating this, in particular the requirement to have both domain.nq and domain-network.nq in the zipfile. Short conclusion is that the domain-network.nq file is no longer required, so can be excluded from the zip file.

At the time of migrating to Github, SSM still required the domain-network artifact as a dependency at build time. This was used when launching SSM to install the default domain models. For this reason, SSM expected domain-network.nq to be available on the classpath, so it could be loaded in as a resource. We also needed the domain-network model available as a zipfile for installing as a knowledgebase bundle (using the domain.nq file, etc). So at this point, we added the domain-network.nq into the build.

Subsequently the need for this dependency was removed, via the refactoring in issue #10. So there is no longer any requirement for domain-network.nq on the classpath, etc. Knowledgebases are installed either at first launch (using domain.nq), or can be uploaded manually, as before. I tested that domain-network.nq  could safely be removed from any uploaded zipfile.

What I wanted to be sure of was that the requirement for the additional domain-network.nq  file had gone away. Certainly I had no problems with the latest dev code. To be sure, I reverted back to a previous version of system-modeller before the #10 branch was merged in. Using this older version, I could reproduce an error when the domain-network.nq  was not present in the dependency:

![image](https://github.com/Spyderisk/system-modeller/assets/50845365/c2f368c5-eacb-486a-809c-075b338ddec9)

Could @mike1813 or @scp93ch fix the domain-network project build script to not include the domain-network.nq in the zipfile? i.e. we need an issue on this.

I will continue to investigate how we can install a zip bundle containing an arbitrary .nq filename (not explictly domain.nq)..

**Comment by kenmeacham** [comment]
It turns out that there is no longer any specific requirement that the domain model file in the zip bundle is called domain.nq. In fact it can be any name you wish, as long as it has a .nq extension. This implies that the domain-network package could contain any reasonable name for its domain file, e.g. domain.nq, domain-network.nq, etc. Whatever is most convenient really.

**Comment by kenmeacham** [comment]
> Ideal: the zip bundle must contain an icons folder plus one NQ and one JSON file, with arbitrary filenames ending '.nq' and '.json', respectively

So we do support arbitrary filenames for the .nq, but currently expect icon-mapping.json. Does this really need an arbitrary filename?

Otherwise I think this issue could be closed. See what you think..


Please try with any zip bundle that you prefer and confirm that this works OK.

**Comment by kenmeacham** [comment]
@mike1813 - can you test this yourself using a single domain file with arbitrary name, etc?

I think this issue can then be closed

**Comment by mike1813** [comment]
Confirmed that in my installation (which dates back to sometime in August 2023), this problem is no longer present.

I also checked the domain-network project. There was no issue raised on changes to the deployable package structure, but the CI pipeline was updated by @scp93ch. However, this was only done on branch '6a'.

I created a new issue there #[76](https://github.com/Spyderisk/domain-network/issues/76), assigned to @scp93ch, asking him to make the new CI pipeline files available also in branch 'master', which is supposed to contain the CI pipeline specification only.

The issue here can be closed.
